mypy/checker.py

```diff
@@ -1715,15 +1797,32 @@ def check_incompatible_property_override(self, e: Decorator) -> None:
 
     def visit_with_stmt(self, s: WithStmt) -> Type:
         echk = self.expr_checker
+        if s.is_async:
+            m_enter = '__aenter__'
+            m_exit = '__aexit__'
+        else:
+            m_enter = '__enter__'
+            m_exit = '__exit__'
         for expr, target in zip(s.expr, s.target):
             ctx = self.accept(expr)
-            enter = echk.analyze_external_member_access('__enter__', ctx, expr)
+            enter = echk.analyze_external_member_access(m_enter, ctx, expr)
             obj = echk.check_call(enter, [], [], expr)[0]
+            if s.is_async:
+                self.check_subtype(obj, self.named_type('typing.Awaitable'), expr)
             if target:
+                if s.is_async:
+                    # XXX TODO What if it's a subclass of Awaitable?
+                    if (isinstance(obj, Instance) and len(obj.args) == 1
+                            and obj.type.fullname() == 'typing.Awaitable'):
+                        obj = obj.args[0]
+                    else:
+                        obj = AnyType()
```

Comment:
Good point. I don't know how to do that, but the same thought had occurred to me. I looked for other places that do this, typically it seems to translate things to some method call and analyzes that. E.g. "for x in blah" seems to analyze `blah.__iter__().__next__()`. Let me see if I can get that to work. It looks like for `await blah` we should analyze `blah.__await__()` and then call `get_generator_return_type()` on the result.